src/hoi4_world/states/hoi4_states_converter.cpp

```diff
@@ -426,6 +426,45 @@ std::tuple<int, int, int> ConvertIndustry(const float& total_factories,
 }
 
 
+std::tuple<std::optional<int>, std::optional<int>> DetermineNavalBase(const std::set<int>& hoi4_provinces,
+    const std::map<std::string, std::string>& significant_vic3_provinces,
+    const hoi4::CoastalProvinces& coastal_provinces,
+    const mappers::Hoi4ToVic3ProvinceMapping& hoi4_to_vic3_province_mappings)
+{
+   for (const auto& hoi4_province: hoi4_provinces)
+   {
+      if (!coastal_provinces.IsProvinceCoastal(hoi4_province))
+      {
+         continue;
+      }
+
+      const auto province_mapping = hoi4_to_vic3_province_mappings.find(hoi4_province);
+      if (province_mapping == hoi4_to_vic3_province_mappings.end())
+      {
+         continue;
+      }
+
+      if (std::ranges::none_of(province_mapping->second,
+              [significant_vic3_provinces](const std::string& vic3_province) {
+                 if (const auto& significant_province = significant_vic3_provinces.find(vic3_province);
+                     significant_province != significant_vic3_provinces.end())
+                 {
+                    return significant_province->second == "port";
+                 }
+
+                 return false;
+              }))
+      {
+         continue;
+      }
+
+      return {hoi4_province, 1};
```

Comment:
You should make a TODO to somehow transfer actual naval base level from building level (goes up to 60 I believe, at max tech) to hoi4 base level 1-10. Divide by 6 or something. Would require pinging the state itself.